lib/psabpf_pipeline.c

```diff
@@ -57,6 +62,68 @@ static int open_prog_by_name(psabpf_context_t *ctx, const char *prog)
     return bpf_obj_get(pinned_file);  // error in errno
 }
 
+static int tc_create_hook(int ifindex, const char *interface)
+{
+    DECLARE_LIBBPF_OPTS(bpf_tc_hook, hook,
+                        .ifindex = ifindex,
+                        .attach_point = BPF_TC_INGRESS | BPF_TC_EGRESS);
+
+    int ret = NO_ERROR;
+    if (bpf_tc_hook_create(&hook) != 0) {
+        ret = errno;
+        fprintf(stderr, "failed to create TC hook for interface %s: %s\n", interface, strerror(ret));
+    }
+
+    return ret;
+}
+
+static int tc_attach_prog(psabpf_context_t *ctx, const char *prog, int ifindex, enum bpf_tc_attach_point hook_point, const char *interface)
+{
+    int ret = NO_ERROR;
+    int fd = open_prog_by_name(ctx, prog);
+    if (fd < 0) {
+        ret = errno;
+        if (ret == ENOENT && hook_point == BPF_TC_EGRESS) {
+            fprintf(stderr, "skipping empty egress program...\n");
+            return NO_ERROR;
+        }
+
+        fprintf(stderr, "failed to open program %s: %s\n", prog, strerror(ret));
+        return ret;
+    }
+
+    DECLARE_LIBBPF_OPTS(bpf_tc_hook, hook,
+                        .ifindex = ifindex,
+                        .attach_point = hook_point);
+    DECLARE_LIBBPF_OPTS(bpf_tc_opts, opts,
+                        .prog_fd = fd);
+
+    if (bpf_tc_attach(&hook, &opts) != 0) {
+        ret = errno;
+        fprintf(stderr, "failed to attach bpf program to interface %s: %s\n", interface, strerror(ret));
+    }
+    close(fd);
```

Comment:
I'd prefer to enclose this with the `goto` section and use `goto` inside the above if block. This is more safe if a user will modify this function in future. With the current approach the prog fd might remain unclosed if the developer wouldn't consider it.